testutil/testproxy/relayerproxy.go

```diff
@@ -0,0 +1,323 @@
+			supplierEndpoints,
+		)
+	}
+}
+
+func WithDefaultSessionSupplier(
+	supplierKeyName string,
+	appPrivateKey *secp256k1.PrivKey,
+) func(*TestBehavior) {
+	return func(test *TestBehavior) {
+		appAddress := GetAddressFromPrivateKey(test, appPrivateKey)
+
+		sessionSuppliers := []string{}
+
+		if supplierKeyName != "" {
+			var keyring keyringtypes.Keyring
+			err := depinject.Inject(test.Deps, &keyring)
+			require.NoError(test.t, err)
+
+			supplierAccount, err := keyring.Key(supplierKeyName)
+			require.NoError(test.t, err)
+
+			supplierAccAddress, err := supplierAccount.GetAddress()
+			require.NoError(test.t, err)
+
+			supplierAddress := supplierAccAddress.String()
+			sessionSuppliers = append(sessionSuppliers, supplierAddress)
+		}
+
+		testqueryclients.AddToExistingSessions(
+			test.t,
+			appAddress,
+			"service1",
+			1,
+			sessionSuppliers,
+		)
+	}
+}
+
+// MarshalAndSend marshals the request and sends it to the provided service
+func MarshalAndSend(
+	test *TestBehavior,
+	url string,
+	request *servicetypes.RelayRequest,
+) (errCode int32, errorMessage string) {
+	reqBz, err := request.Marshal()
+	require.NoError(test.t, err)
+
+	reader := io.NopCloser(bytes.NewReader(reqBz))
+	res, err := http.DefaultClient.Post(url, "application/json", reader)
+	require.NoError(test.t, err)
+	require.NotNil(test.t, res)
+
+	return GetRelayResponseError(test.t, res)
+}
+
+// GetRelayResponseError returns the error code and message from the relay response
+// if the response is not an error, it returns 0, ""
+func GetRelayResponseError(t *testing.T, res *http.Response) (errCode int32, errMsg string) {
+	responseBody, err := io.ReadAll(res.Body)
+	require.NoError(t, err)
+
+	relayResponse := &servicetypes.RelayResponse{}
+	err = relayResponse.Unmarshal(responseBody)
+	if err != nil {
+		return 0, "cannot unmarshal response body"
+	}
+
+	var payload JSONRpcErrorReply
+	err = json.Unmarshal(relayResponse.Payload, &payload)
+	if err != nil {
+		return 0, "cannot unmarshal response payload"
+	}
+
+	if payload.Error == nil {
+		return 0, ""
+	}
+
+	return payload.Error.Code, payload.Error.Message
+}
+
+// GetRelayResponseResult crafts a ring signer for test purposes and uses it to
+// sign the relay request
+func GetApplicationRingSignature(
+	t *testing.T,
+	req *servicetypes.RelayRequest,
+	appPrivateKey *secp256k1.PrivKey,
+) []byte {
+	publicKey := appPrivateKey.PubKey()
+	curve := ring_secp256k1.NewCurve()
+
+	point, err := curve.DecodeToPoint(publicKey.Bytes())
+	require.NoError(t, err)
+
+	// At least two points are required to create a ring signer
+	points := []ringtypes.Point{point, point}
+	pointsRing, err := ring.NewFixedKeyRingFromPublicKeys(curve, points)
+	require.NoError(t, err)
+
+	scalar, err := curve.DecodeToScalar(appPrivateKey.Bytes())
+	require.NoError(t, err)
+
+	signer := signer.NewRingSigner(pointsRing, scalar)
+
+	signableBz, err := req.GetSignableBytes()
+	require.NoError(t, err)
+
+	hash := crypto.Sha256(signableBz)
```

Comment:
We should be signing `signableBz`. Why is there an extra hash here?

Reply:
`req.GetSignableBytes` just marshals the `RelayRequest` (excluding the signature field). Which means it results in an arbitrary length byte slice. We need to have a 32 bytes length content to sign.

Reply:
That's a great point but it means we have a large oversight in a fundamental component of our codebase. `GetSignableBytes` is going to be very important in:
- Signing relays
- Verifying relays
- Claim & Proof lifecycle
- Indexers
- Challenges
- Etc...

The fact that this logic (which is key to state transitions) is not centralized is a big problem IMO.

I think we need to create a new PR ASAP (and merge it in before this one) that updates `GetSignableBytes` to return a constant length string.

1. Wdyt?
2. Would you be able to do that?

cc @h5law